tests/test.c

```diff
@@ -24,134 +24,48 @@
-
-static char opt_badstring[5];
-string_cmd("badstring", opt_badstring);
-
-static unsigned long opt_ulong;
-ulong_cmd("integer", opt_ulong);
-
-static bool opt_bool = 0;
-bool_cmd("boolean", opt_bool);
+static const char opt_test_delims[] = ",";
+#include <cmdline.h>
 
-static bool opt_booltwo = 0;
-bool_cmd("booleantwo", opt_booltwo);
+static char opt_tests[MAX_OPT_TESTS_LEN];
+string_cmd("tests", opt_tests);
 
-static char memmove_string[4];
-static char range_string[] = "123456";
-static char *src, *dst;
+static test_fn *get_next_test(void) {
+    static char *opt = opt_tests;
+    char *test_name;
+    test_fn *test = NULL;
 
-static void cpu_freq_expect(const char *cpu_str, uint64_t expectation) {
-    uint64_t result = get_cpu_freq(cpu_str);
-    if (result != expectation) {
-        printk("Could not parse cpu frequency from string '%s'\n", cpu_str);
-        printk("Expectation vs. result: %llu vs. %llu\n", expectation, result);
-        BUG();
+    test_name = strtok(opt, opt_test_delims);
+    opt = NULL;
+    if (test_name) {
+        test = symbol_address(test_name);
+        if (!test)
+            panic("Symbol for test %s not found\n", test_name);
     }
-
-    printk("Got CPU string '%s' and frequency '%llu'\n", cpu_str, result);
-    return;
+    return test;
 }
 
-static int __user_text func(void *arg) { return 0; }
-
-static int ktf_unit_tests(void) {
-    usermode_call(func, NULL);
-
-    printk("\nLet the UNITTESTs begin\n");
-    printk("Commandline parsing: %s\n", kernel_cmdline);
-
-    if (strcmp(opt_string, "foo")) {
-        printk("String parameter opt_string != foo: %s\n", opt_string);
-        BUG();
-    }
-    else {
-        printk("String parameter parsing works!\n");
-    }
-
-    if (strcmp(opt_badstring, "tool")) {
-        printk("String parameter opt_badstring != tool: %s\n", opt_badstring);
-        BUG();
-    }
-    else {
-        printk("String parameter parsing works!\n");
-    }
-
-    if (opt_ulong != 42) {
-        printk("Integer parameter opt_ulong != 42: %d\n", opt_ulong);
-        BUG();
-    }
-    else {
-        printk("Integer parameter parsing works!\n");
-    }
+void test_main(void) {
+    test_fn *test;
+    unsigned n = 0;
 
-    if (!opt_bool || !opt_booltwo) {
-        printk("Boolean parameter opt_bool != true: %d\n", opt_bool);
-        printk("Boolean parameter opt_booltwo != true: %d\n", opt_booltwo);
-        BUG();
-    }
-    else {
-        printk("Boolean parameter parsing works!\n");
-    }
+    printk("\nRunning tests\n");
 
-    printk("\nMemmove testing:\n");
-    (void) memmove(memmove_string, opt_string, sizeof(opt_string));
-    if (!strcmp(memmove_string, opt_string)) {
-        printk("Moving around memory works!\n");
-    }
-    else {
-        printk("Memmove'ing did not work: %s (%p) != %s (%p)\n", memmove_string,
-               memmove_string, opt_string, opt_string);
-    }
+    while ((test = get_next_test()) != NULL) {
+        int rc;
```

Comment:
it would be helpful to obtain test's name also, in order to highlight beginning of each test as well as clearly mark which test returned what value.